File: scripts/repo-update-log.js
#!/usr/bin/env node
"use strict";

const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const {
  DEFAULT_IGNORE_DIRS,
  findProjectRoot,
  normalizePath,
  resolveAgentAccessControl,
  resolveUpdateLogPaths
} = require("./project-source-resolver");

const RUNTIME = {
  root: "",
  policyPath: "",
  eventsLogFile: "",
  sessionLogFile: "",
  stateFile: "",
  ignoredPaths: []
};

function nowIso() {
  return new Date().toISOString();
}

function createEventId() {
  if (typeof crypto.randomUUID === "function") {
    return `evt_${crypto.randomUUID()}`;
  }
  return `evt_${crypto.randomBytes(16).toString("hex")}`;
}

function printHelpAndExit(code) {
  const helpText = [
    "repo-update-log",
    "",
    "Usage:",
    "  npm run updates:scan -- [options]",
    "  npm run updates:watch -- [options]",
    "",
    "Commands:",
    "  scan   Compare current project state to previous snapshot and log file changes",
    "  watch  Watch project scope for real-time file update events",
    "",
    "Options:",
    "  --actor <id>             Actor tag for events (default: system)",
    "  --scope <text>           Scope/reason tag for scan mode",
    "  --session-id <id>        Explicit session id (watch mode)",
    "  --help                   Show help"
  ].join("\n");
  process.stdout.write(`${helpText}\n`);
  process.exit(code);
}

function parseArgs(argv) {
  if (!argv.length || argv.includes("--help") || argv.includes("-h")) {
    printHelpAndExit(0);
  }

  const args = {
    command: String(argv[0] || "")
      .trim()
      .toLowerCase(),
    actor: "system",
    scope: "",
    sessionId: ""
  };

  for (let index = 1; index < argv.length; index += 1) {
    const token = argv[index];

    if (token === "--actor") {
      args.actor = String(argv[index + 1] || "").trim() || "system";
      index += 1;
      continue;
    }

    if (token === "--scope") {
      args.scope = String(argv[index + 1] || "").trim();
      index += 1;
      continue;
    }

    if (token === "--session-id") {
      args.sessionId = String(argv[index + 1] || "").trim();
      index += 1;
      continue;
    }

    throw new Error(`unknown argument: ${token}`);
  }

  if (!["scan", "watch"].includes(args.command)) {
    throw new Error("first argument must be one of: scan, watch");
  }

  return args;
}

function ensureDir(dirPath) {
  fs.mkdirSync(dirPath, { recursive: true });
}

function unique(values) {
  return [...new Set(values)];
}

function configureRuntime() {
  let root = findProjectRoot(process.cwd());
  let policyPath = "";
  let updateLogPaths = null;

  try {
    const resolved = resolveAgentAccessControl(process.cwd());
    root = resolved.root;
    policyPath = resolved.policyPath;
    updateLogPaths = resolveUpdateLogPaths(root, policyPath, resolved.policy);
  } catch {
    const fallbackDir = path.join(root, "data", "logs", "change-log");
    updateLogPaths = {
      eventsFile: path.join(fallbackDir, "update_events.ndjson"),
      sessionsFile: path.join(fallbackDir, "sessions.ndjson"),
      stateFile: path.join(fallbackDir, "state_snapshot.json")
    };
  }

  RUNTIME.root = root;
  RUNTIME.policyPath = policyPath;
  RUNTIME.eventsLogFile = updateLogPaths.eventsFile;
  RUNTIME.sessionLogFile = updateLogPaths.sessionsFile;
  RUNTIME.stateFile = updateLogPaths.stateFile;

  const derivedIgnored = [
    ...DEFAULT_IGNORE_DIRS,
    normalizePath(path.relative(root, path.dirname(updateLogPaths.eventsFile))),
    normalizePath(path.relative(root, path.dirname(updateLogPaths.sessionsFile))),
    normalizePath(path.relative(root, path.dirname(updateLogPaths.stateFile)))
  ].filter((entry) => Boolean(entry) && entry !== ".");

  RUNTIME.ignoredPaths = unique(derivedIgnored);
}

function ensureLogPaths() {
  ensureDir(path.dirname(RUNTIME.eventsLogFile));
  ensureDir(path.dirname(RUNTIME.sessionLogFile));
  if (!fs.existsSync(RUNTIME.eventsLogFile)) {
    fs.writeFileSync(RUNTIME.eventsLogFile, "", "utf8");
  }
  if (!fs.existsSync(RUNTIME.sessionLogFile)) {
    fs.writeFileSync(RUNTIME.sessionLogFile, "", "utf8");
  }
}

function toRelativePath(filePath) {
  return normalizePath(path.relative(RUNTIME.root, filePath));
}

function isIgnored(relativePath) {
  const rel = normalizePath(relativePath || "");
  if (!rel || rel === ".") {
    return false;
  }
  return RUNTIME.ignoredPaths.some((ignorePath) => rel === ignorePath || rel.startsWith(`${ignorePath}/`));
}

function createFileHash(filePath) {
  const content = fs.readFileSync(filePath);
  return crypto.createHash("sha256").update(content).digest("hex");
}

function fileSnapshot(filePath) {
  if (!fs.existsSync(filePath)) {
    return null;
  }
  const stat = fs.statSync(filePath);
  if (!stat.isFile()) {
    return null;
  }
  return {
    size_bytes: stat.size,
    mtime_ms: Number(stat.mtimeMs.toFixed(3)),
    hash: createFileHash(filePath)
  };
}

function scanFileState() {
  const state = {};
  const stack = [RUNTIME.root];

  while (stack.length > 0) {
    const currentDir = stack.pop();
    const entries = fs.readdirSync(currentDir, { withFileTypes: true });

    entries.forEach((entry) => {
      const abs = path.join(currentDir, entry.name);
      const rel = toRelativePath(abs);
      if (isIgnored(rel)) {
        return;
      }
      if (entry.isDirectory()) {
        stack.push(abs);
        return;
      }
      if (!entry.isFile()) {
        return;
      }
      const snapshot = fileSnapshot(abs);
      if (snapshot) {
        state[rel] = snapshot;
      }
    });
  }

  return state;
}

function readState() {
  if (!fs.existsSync(RUNTIME.stateFile)) {
    return {
      files: {},
      updated_at: null
    };
  }
  try {
    const parsed = JSON.parse(fs.readFileSync(RUNTIME.stateFile, "utf8"));
    return {
      files: parsed && parsed.files ? parsed.files : {},
      updated_at: parsed && parsed.updated_at ? parsed.updated_at : null
    };
  } catch {
    return {
      files: {},
      updated_at: null
    };
  }
}

function writeState(fileMap) {
  ensureDir(path.dirname(RUNTIME.stateFile));
  fs.writeFileSync(
    RUNTIME.stateFile,
    `${JSON.stringify(
      {
        updated_at: nowIso(),
        file_count: Object.keys(fileMap).length,
        files: fileMap
      },
      null,
      2
    )}\n`,
    "utf8"
  );
}

function appendNdjson(filePath, payload) {
  ensureDir(path.dirname(filePath));
  fs.appendFileSync(filePath, `${JSON.stringify(payload)}\n`, "utf8");
}

function classifyDiffs(previousFiles, currentFiles) {
  const allPaths = [...new Set([...Object.keys(previousFiles), ...Object.keys(currentFiles)])].sort();
  const diffs = [];

  allPaths.forEach((relPath) => {
    const before = previousFiles[relPath] || null;
    const after = currentFiles[relPath] || null;

    if (!before && after) {
      diffs.push({
        change_type: "file_created",
        path: relPath,
        before,
        after
      });
      return;
    }

    if (before && !after) {
      diffs.push({
        change_type: "file_deleted",
        path: relPath,
        before,
        after
      });
      return;
    }

    if (
      before &&
      after &&
      (before.hash !== after.hash || before.size_bytes !== after.size_bytes || before.mtime_ms !== after.mtime_ms)
    ) {
      diffs.push({
        change_type: "file_updated",
        path: relPath,
        before,
        after
      });
    }
  });

  return diffs;
}

function runScanMode(args) {
  ensureLogPaths();
  const startedAtMs = Date.now();
  const startedAt = nowIso();

  const previousState = readState();
  const currentFiles = scanFileState();
  const diffs = classifyDiffs(previousState.files, currentFiles);

  const eventRows = diffs.map((diff, index) => ({
    event_id: createEventId(),
    event_at: nowIso(),
    actor: args.actor,
    source: "scan",
    scope: args.scope || "unspecified",
    change_type: diff.change_type,
    path: diff.path,
    elapsed_ms_from_scan_start: Date.now() - startedAtMs,
    before: diff.before,
    after: diff.after,
    diff_index: index + 1
  }));

  eventRows.forEach((row) => appendNdjson(RUNTIME.eventsLogFile, row));
  writeState(currentFiles);

  const finishedAt = nowIso();
  const durationMs = Date.now() - startedAtMs;
  const summary = {
    session_id: createEventId().replace(/^evt_/, "scan_"),
    mode: "scan",
    actor: args.actor,
    scope: args.scope || "unspecified",
    started_at: startedAt,
    finished_at: finishedAt,
    duration_ms: durationMs,
    changes_total: eventRows.length,
    changed_paths: eventRows.map((row) => row.path)
  };
  appendNdjson(RUNTIME.sessionLogFile, summary);

  process.stdout.write(`${JSON.stringify(summary, null, 2)}\n`);
}

function buildDirectoryList(baseDirPath) {
  const dirs = [];
  const stack = [baseDirPath];

  while (stack.length > 0) {
    const current = stack.pop();
    const relCurrent = toRelativePath(current);
    if (isIgnored(relCurrent)) {
      continue;
    }
    dirs.push(current);
    const entries = fs.readdirSync(current, { withFileTypes: true });
    entries.forEach((entry) => {
      if (!entry.isDirectory()) {
        return;
      }
      const abs = path.join(current, entry.name);
      const rel = toRelativePath(abs);
      if (!isIgnored(rel)) {
        stack.push(abs);
      }
    });
  }

  return dirs;
}

function runWatchMode(args) {
  ensureLogPaths();
  const startedAtMs = Date.now();
  const startedAt = nowIso();
  const sessionId = args.sessionId || `watch_${createEventId().replace(/^evt_/, "")}`;

  const knownState = scanFileState();
  writeState(knownState);

  const watchers = new Map();
  const pendingByPath = new Map();

  function saveKnownState() {
    writeState(knownState);
  }

  function recordEvent(changeType, relPath, before, after) {
    const event = {
      event_id: createEventId(),
      event_at: nowIso(),
      actor: args.actor,
      source: "watch",
      session_id: sessionId,
      change_type: changeType,
      path: relPath,
      elapsed_ms_from_session_start: Date.now() - startedAtMs,
      before: before || null,
      after: after || null
    };
    appendNdjson(RUNTIME.eventsLogFile, event);
  }

  function removeWatcherByPrefix(absPathPrefix) {
    const normalized = path.resolve(absPathPrefix);
    Array.from(watchers.keys()).forEach((watchedDir) => {
      if (watchedDir === normalized || watchedDir.startsWith(`${normalized}${path.sep}`)) {
        const watcher = watchers.get(watchedDir);
        if (watcher) {
          watcher.close();
        }
        watchers.delete(watchedDir);
      }
    });
  }

  function syncPath(relativePath) {
    if (!relativePath || isIgnored(relativePath)) {
      return;
    }

    const abs = path.join(RUNTIME.root, relativePath);
    const before = knownState[relativePath] || null;

    if (!fs.existsSync(abs)) {
      if (before) {
        delete knownState[relativePath];
        recordEvent("file_deleted", relativePath, before, null);
        saveKnownState();
      }
      return;
    }

    const stat = fs.statSync(abs);
    if (stat.isDirectory()) {
      const dirs = buildDirectoryList(abs);
      dirs.forEach((dirPath) => addWatcher(dirPath));
      return;
    }

    if (!stat.isFile()) {
      return;
    }

    const after = fileSnapshot(abs);
    if (!before && after) {
      knownState[relativePath] = after;
      recordEvent("file_created", relativePath, null, after);
      saveKnownState();
      return;
    }

    if (
      before &&
      after &&
      (before.hash !== after.hash || before.size_bytes !== after.size_bytes || before.mtime_ms !== after.mtime_ms)
    ) {
      knownState[relativePath] = after;
      recordEvent("file_updated", relativePath, before, after);
      saveKnownState();
    }
  }

  function scheduleSync(relativePath) {
    if (!relativePath || isIgnored(relativePath)) {
      return;
    }
    const existing = pendingByPath.get(relativePath);
    if (existing) {
      clearTimeout(existing);
    }
    const timer = setTimeout(() => {
      pendingByPath.delete(relativePath);
      syncPath(relativePath);
    }, 75);
    pendingByPath.set(relativePath, timer);
  }

  function addWatcher(directoryPath) {
    const absDir = path.resolve(directoryPath);
    if (watchers.has(absDir)) {
      return;
    }
    const relDir = toRelativePath(absDir);
    if (isIgnored(relDir)) {
      return;
    }
    let watcher = null;
    try {
      watcher = fs.watch(absDir, { persistent: true }, (eventType, filename) => {
        const name = String(filename || "").trim();
        if (!name) {
          return;
        }
        const targetAbs = path.resolve(absDir, name);
        const targetRel = toRelativePath(targetAbs);
        if (eventType === "rename" && !fs.existsSync(targetAbs)) {
          removeWatcherByPrefix(targetAbs);
        }
        scheduleSync(targetRel);
      });
    } catch {
      return;
    }

    watcher.on("error", () => {
      watchers.delete(absDir);
    });
    watchers.set(absDir, watcher);
  }

  const startupDirs = buildDirectoryList(RUNTIME.root);
  startupDirs.forEach((dirPath) => addWatcher(dirPath));

  const sessionStartPayload = {
    session_id: sessionId,
    mode: "watch",
    actor: args.actor,
    started_at: startedAt,
    status: "started",
    watcher_count: watchers.size
  };
  appendNdjson(RUNTIME.sessionLogFile, sessionStartPayload);

  function shutdown(signalName) {
    pendingByPath.forEach((timer) => clearTimeout(timer));
    watchers.forEach((watcher) => watcher.close());
    watchers.clear();
    saveKnownState();
    appendNdjson(RUNTIME.sessionLogFile, {
      session_id: sessionId,
      mode: "watch",
      actor: args.actor,
      started_at: startedAt,
      finished_at: nowIso(),
      duration_ms: Date.now() - startedAtMs,
      status: "stopped",
      signal: signalName
    });
    process.exit(0);
  }

  process.on("SIGINT", () => shutdown("SIGINT"));
  process.on("SIGTERM", () => shutdown("SIGTERM"));
  process.on("exit", () => {
    if (watchers.size > 0) {
      watchers.forEach((watcher) => watcher.close());
      watchers.clear();
    }
  });

  process.stdout.write(
    `${JSON.stringify(
      {
        ok: true,
        mode: "watch",
        session_id: sessionId,
        actor: args.actor,
        project_root: RUNTIME.root,
        policy_file: RUNTIME.policyPath || null,
        watcher_count: watchers.size,
        events_file: path.relative(RUNTIME.root, RUNTIME.eventsLogFile),
        sessions_file: path.relative(RUNTIME.root, RUNTIME.sessionLogFile)
      },
      null,
      2
    )}\n`
  );
}

function run() {
  configureRuntime();
  const args = parseArgs(process.argv.slice(2));
  if (args.command === "scan") {
    runScanMode(args);
    return;
  }
  runWatchMode(args);
}

try {
  run();
} catch (error) {
  process.stderr.write(`repo-update-log failed: ${error.message}\n`);
  process.exit(1);
}

File: scripts/polyglot/equivalents/python/_native/repo_update_log.py
#!/usr/bin/env python3
"""Native Python implementation for scripts/repo-update-log.js."""

from __future__ import annotations

import hashlib
import json
import os
import pathlib
import signal
import sys
import time
import uuid
from datetime import UTC, datetime
from typing import Any

from _common import normalize_path, write_text_file_robust
from project_source_resolver_native import (
    DEFAULT_IGNORE_DIRS,
    find_project_root,
    resolve_agent_access_control,
    resolve_update_log_paths,
)

RUNTIME: dict[str, Any] = {
    "root": pathlib.Path(),
    "policy_path": pathlib.Path(),
    "events_log_file": pathlib.Path(),
    "session_log_file": pathlib.Path(),
    "state_file": pathlib.Path(),
    "ignored_paths": [],
}


def now_iso() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def create_event_id() -> str:
    return f"evt_{uuid.uuid4()}"


def print_help_and_exit(code: int) -> None:
    help_text = "\n".join(
        [
            "repo-update-log",
            "",
            "Usage:",
            "  npm run updates:scan -- [options]",
            "  npm run updates:watch -- [options]",
            "",
            "Commands:",
            "  scan   Compare current project state to previous snapshot and log file changes",
            "  watch  Watch project scope for real-time file update events",
            "",
            "Options:",
            "  --actor <id>             Actor tag for events (default: system)",
            "  --scope <text>           Scope/reason tag for scan mode",
            "  --session-id <id>        Explicit session id (watch mode)",
            "  --help                   Show help",
        ]
    )
    sys.stdout.write(f"{help_text}\n")
    raise SystemExit(code)


def parse_args(argv: list[str]) -> dict[str, str]:
    if not argv or "--help" in argv or "-h" in argv:
        print_help_and_exit(0)
    args = {"command": str(argv[0] or "").strip().lower(), "actor": "system", "scope": "", "session_id": ""}
    index = 1
    while index < len(argv):
        token = str(argv[index] or "")
        if token == "--actor":
            args["actor"] = str(argv[index + 1] or "").strip() or "system"
            index += 2
            continue
        if token == "--scope":
            args["scope"] = str(argv[index + 1] or "").strip()
            index += 2
            continue
        if token == "--session-id":
            args["session_id"] = str(argv[index + 1] or "").strip()
            index += 2
            continue
        raise RuntimeError(f"unknown argument: {token}")
    if args["command"] not in {"scan", "watch"}:
        raise RuntimeError("first argument must be one of: scan, watch")
    return args


def ensure_dir(dir_path: pathlib.Path) -> None:
    dir_path.mkdir(parents=True, exist_ok=True)


def unique(values: list[str]) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for value in values:
        key = str(value or "")
        if not key or key in seen:
            continue
        seen.add(key)
        out.append(key)
    return out


def configure_runtime() -> None:
    root = find_project_root(pathlib.Path.cwd())
    policy_path = pathlib.Path()
    update_log_paths: dict[str, pathlib.Path]
    try:
        resolved = resolve_agent_access_control(root)
        root = resolved["root"]
        policy_path = resolved["policy_path"]
        update_log_paths = resolve_update_log_paths(root, policy_path, resolved["policy"])
    except Exception:
        fallback_dir = root / "data" / "logs" / "change-log"
        update_log_paths = {
            "events_file": fallback_dir / "update_events.ndjson",
            "sessions_file": fallback_dir / "sessions.ndjson",
            "state_file": fallback_dir / "state_snapshot.json",
        }

    RUNTIME["root"] = root
    RUNTIME["policy_path"] = policy_path
    RUNTIME["events_log_file"] = update_log_paths["events_file"]
    RUNTIME["session_log_file"] = update_log_paths["sessions_file"]
    RUNTIME["state_file"] = update_log_paths["state_file"]
    derived_ignored = [
        *DEFAULT_IGNORE_DIRS,
        normalize_path(update_log_paths["events_file"].parent.relative_to(root)),
        normalize_path(update_log_paths["sessions_file"].parent.relative_to(root)),
        normalize_path(update_log_paths["state_file"].parent.relative_to(root)),
    ]
    RUNTIME["ignored_paths"] = [entry for entry in unique(derived_ignored) if entry and entry != "."]


def ensure_log_paths() -> None:
    ensure_dir(RUNTIME["events_log_file"].parent)
    ensure_dir(RUNTIME["session_log_file"].parent)
    if not RUNTIME["events_log_file"].exists():
        RUNTIME["events_log_file"].write_text("", encoding="utf8")
    if not RUNTIME["session_log_file"].exists():
        RUNTIME["session_log_file"].write_text("", encoding="utf8")


def to_relative_path(file_path: pathlib.Path) -> str:
    return normalize_path(file_path.relative_to(RUNTIME["root"]))


def is_ignored(relative_path: str) -> bool:
    rel = normalize_path(relative_path or "")
    if not rel or rel == ".":
        return False
    return any(rel == ignore_path or rel.startswith(f"{ignore_path}/") for ignore_path in RUNTIME["ignored_paths"])


def create_file_hash(file_path: pathlib.Path) -> str:
    return hashlib.sha256(file_path.read_bytes()).hexdigest()


def file_snapshot(file_path: pathlib.Path) -> dict[str, Any] | None:
    if not file_path.exists() or not file_path.is_file():
        return None
    stat = file_path.stat()
    return {
        "size_bytes": stat.st_size,
        "mtime_ms": round(stat.st_mtime_ns / 1_000_000.0, 3),
        "hash": create_file_hash(file_path),
    }


def scan_file_state() -> dict[str, dict[str, Any]]:
    state: dict[str, dict[str, Any]] = {}
    stack = [RUNTIME["root"]]
    while stack:
        current_dir = stack.pop()
        try:
            entries = list(current_dir.iterdir())
        except PermissionError:
            continue
        for entry in entries:
            rel = to_relative_path(entry)
            if is_ignored(rel):
                continue
            if entry.is_dir():
                stack.append(entry)
                continue
            if not entry.is_file():
                continue
            snapshot = file_snapshot(entry)
            if snapshot is not None:
                state[rel] = snapshot
    return state


def read_state() -> dict[str, Any]:
    state_file = RUNTIME["state_file"]
    if not state_file.exists():
        return {"files": {}, "updated_at": None}
    try:
        parsed = json.loads(state_file.read_text(encoding="utf8"))
    except Exception:
        return {"files": {}, "updated_at": None}
    return {
        "files": parsed.get("files") if isinstance(parsed.get("files"), dict) else {},
        "updated_at": parsed.get("updated_at"),
    }


def write_state(file_map: dict[str, Any]) -> None:
    payload = {
        "updated_at": now_iso(),
        "file_count": len(file_map),
        "files": file_map,
    }
    write_text_file_robust(RUNTIME["state_file"], f"{json.dumps(payload, indent=2)}\n", atomic=False)


def append_ndjson(file_path: pathlib.Path, payload: dict[str, Any]) -> None:
    ensure_dir(file_path.parent)
    with file_path.open("a", encoding="utf8") as handle:
        handle.write(f"{json.dumps(payload)}\n")


def classify_diffs(previous_files: dict[str, Any], current_files: dict[str, Any]) -> list[dict[str, Any]]:
    all_paths = sorted(set(previous_files.keys()) | set(current_files.keys()))
    diffs: list[dict[str, Any]] = []
    for rel_path in all_paths:
        before = previous_files.get(rel_path)
        after = current_files.get(rel_path)
        if before is None and after is not None:
            diffs.append({"change_type": "file_created", "path": rel_path, "before": None, "after": after})
            continue
        if before is not None and after is None:
            diffs.append({"change_type": "file_deleted", "path": rel_path, "before": before, "after": None})
            continue
        if before != after:
            diffs.append({"change_type": "file_updated", "path": rel_path, "before": before, "after": after})
    return diffs


def run_scan_mode(args: dict[str, str]) -> int:
    ensure_log_paths()
    started_at_ms = int(time.time() * 1000)
    started_at = now_iso()
    previous_state = read_state()
    current_files = scan_file_state()
    diffs = classify_diffs(previous_state.get("files") or {}, current_files)
    event_rows = []
    for index, diff in enumerate(diffs, start=1):
        event_rows.append(
            {
                "event_id": create_event_id(),
                "event_at": now_iso(),
                "actor": args["actor"],
                "source": "scan",
                "scope": args["scope"] or "unspecified",
                "change_type": diff["change_type"],
                "path": diff["path"],
                "elapsed_ms_from_scan_start": int(time.time() * 1000) - started_at_ms,
                "before": diff["before"],
                "after": diff["after"],
                "diff_index": index,
            }
        )
    for row in event_rows:
        append_ndjson(RUNTIME["events_log_file"], row)
    write_state(current_files)
    finished_at = now_iso()
    duration_ms = int(time.time() * 1000) - started_at_ms
    summary = {
        "session_id": create_event_id().replace("evt_", "scan_"),
        "mode": "scan",
        "actor": args["actor"],
        "scope": args["scope"] or "unspecified",
        "started_at": started_at,
        "finished_at": finished_at,
        "duration_ms": duration_ms,
        "changes_total": len(event_rows),
        "changed_paths": [row["path"] for row in event_rows],
    }
    append_ndjson(RUNTIME["session_log_file"], summary)
    sys.stdout.write(f"{json.dumps(summary, indent=2)}\n")
    return 0


def run_watch_mode(args: dict[str, str]) -> int:
    ensure_log_paths()
    started_at_ms = int(time.time() * 1000)
    started_at = now_iso()
    session_id = args["session_id"] or create_event_id().replace("evt_", "watch_")
    known_state = scan_file_state()
    write_state(known_state)
    append_ndjson(
        RUNTIME["session_log_file"],
        {
            "session_id": session_id,
            "mode": "watch",
            "actor": args["actor"],
            "started_at": started_at,
            "status": "started",
            "watcher_count": 1,
        },
    )
    sys.stdout.write(
        f"{json.dumps({'ok': True, 'mode': 'watch', 'session_id': session_id, 'actor': args['actor'], 'project_root': str(RUNTIME['root']), 'policy_file': str(RUNTIME['policy_path']) if str(RUNTIME['policy_path']) else None, 'watcher_count': 1, 'events_file': normalize_path(RUNTIME['events_log_file'].relative_to(RUNTIME['root'])), 'sessions_file': normalize_path(RUNTIME['session_log_file'].relative_to(RUNTIME['root']))}, indent=2)}\n"
    )
    stop_requested = False

    def request_stop(*_: Any) -> None:
        nonlocal stop_requested
        stop_requested = True

    signal.signal(signal.SIGINT, request_stop)
    signal.signal(signal.SIGTERM, request_stop)
    try:
        while not stop_requested:
            time.sleep(0.25)
            current_state = scan_file_state()
            diffs = classify_diffs(known_state, current_state)
            if not diffs:
                continue
            for diff in diffs:
                append_ndjson(
                    RUNTIME["events_log_file"],
                    {
                        "event_id": create_event_id(),
                        "event_at": now_iso(),
                        "actor": args["actor"],
                        "source": "watch",
                        "session_id": session_id,
                        "change_type": diff["change_type"],
                        "path": diff["path"],
                        "elapsed_ms_from_session_start": int(time.time() * 1000) - started_at_ms,
                        "before": diff["before"],
                        "after": diff["after"],
                    },
                )
            known_state = current_state
            write_state(known_state)
    finally:
        write_state(known_state)
        append_ndjson(
            RUNTIME["session_log_file"],
            {
                "session_id": session_id,
                "mode": "watch",
                "actor": args["actor"],
                "started_at": started_at,
                "finished_at": now_iso(),
                "duration_ms": int(time.time() * 1000) - started_at_ms,
                "status": "stopped",
                "signal": "SIGINT" if stop_requested else "exit",
            },
        )
    return 0


def main(argv: list[str] | None = None) -> int:
    configure_runtime()
    args = parse_args(argv or [])
    if args["command"] == "scan":
        return run_scan_mode(args)
    return run_watch_mode(args)


if __name__ == "__main__":
    try:
        raise SystemExit(main(sys.argv[1:]))
    except SystemExit:
        raise
    except Exception as error:
        sys.stderr.write(f"repo-update-log failed: {error}\n")
        raise SystemExit(1)
